brave/src/main/java/brave/propagation/ExtraFieldPropagation.java

```diff
@@ -252,32 +252,55 @@ public static void set(TraceContext context, String name, String value) {
 
     @Override public TraceContext decorate(TraceContext context) {
       TraceContext result = delegate.decorate(context);
-      int extraIndex = indexOfExtra(result.extra());
-      if (extraIndex != -1) {
-        Extra extra = (Extra) result.extra().get(extraIndex);
-        if (!Arrays.equals(extra.fieldNames, fieldNames)) {
-          throw new IllegalStateException(
-              String.format("Mixed name configuration unsupported: found %s, expected %s",
-                  Arrays.asList(extra.fieldNames), Arrays.asList(fieldNames))
-          );
-        }
 
-        // If this extra is unassociated (due to remote extraction),
-        // or it is the same span ID, re-use the instance.
-        if (extra.tryAssociate(context)) return context;
+      // If there's an implicit context while extracting only extra fields, we will have two extras!
+      List<Object> extras = result.extra();
+      int thisExtraIndex = -1, parentExtraIndex = -1;
+      for (int i = 0, length = extras.size(); i < length; i++) {
+        if (extras.get(i) instanceof Extra) {
+          Extra extra = (Extra) result.extra().get(i);
+          if (!Arrays.equals(extra.fieldNames, fieldNames)) {
+            throw new IllegalStateException(
+                String.format("Mixed name configuration unsupported: found %s, expected %s",
+                    Arrays.asList(extra.fieldNames), Arrays.asList(fieldNames))
```

Comment:
I find it rather hard to wrap my head around, so I don't have a lot of comments other than you can use `Arrays.toString(...)` instead of create a complete new list here. But yeah, it's in the exception flow and not in the critical path or something... :-)

Reply:
good hint!